go/oasis-node/cmd/debug/txsource/workload/registration.go

```diff
@@ -0,0 +1,343 @@
+	// consensus stopping as the registered validators wouldn't actually
+	// exist.
+	availableRoles := []node.RolesMask{
+		node.RoleStorageWorker,
+		node.RoleComputeWorker,
+		node.RoleStorageWorker | node.RoleComputeWorker,
+	}
+
+	nodeDesc := node.Node{
+		ID:         nodeIdentity.NodeSigner.Public(),
+		EntityID:   entityID,
+		Expiration: 0,
+		Roles:      availableRoles[rng.Intn(len(availableRoles))],
+		Committee: node.CommitteeInfo{
+			Certificate: nodeIdentity.TLSCertificate.Certificate[0],
+			Addresses: []node.CommitteeAddress{
+				{
+					Certificate: nodeIdentity.TLSCertificate.Certificate[0],
+					Address:     nodeAddr,
+				},
+			},
+		},
+		P2P: node.P2PInfo{
+			ID: nodeIdentity.P2PSigner.Public(),
+			Addresses: []node.Address{
+				nodeAddr,
+			},
+		},
+		Consensus: node.ConsensusInfo{
+			ID: nodeIdentity.ConsensusSigner.Public(),
+			Addresses: []node.ConsensusAddress{
+				{
+					ID:      nodeIdentity.ConsensusSigner.Public(),
+					Address: nodeAddr,
+				},
+			},
+		},
+		Runtimes: []*node.Runtime{
+			{
+				ID: ns,
+			},
+		},
+	}
+	return &nodeDesc
+}
+
+func signNode(identity *identity.Identity, nodeDesc *node.Node) (*node.MultiSignedNode, error) {
+	nodeSigners := []signature.Signer{
+		identity.NodeSigner,
+		identity.P2PSigner,
+		identity.ConsensusSigner,
+		identity.TLSSigner,
+	}
+
+	sigNode, err := node.MultiSignNode(nodeSigners, registry.RegisterNodeSignatureContext, nodeDesc)
+	if err != nil {
+		registryLogger.Error("failed to sign node descriptor",
+			"err", err,
+		)
+		return nil, err
+	}
+
+	return sigNode, nil
+}
+
+func getAccountNonce(ctx context.Context, conn *grpc.ClientConn, pubKey signature.PublicKey) (uint64, error) {
+	stakingClient := staking.NewStakingClient(conn)
+	account, err := stakingClient.AccountInfo(ctx, &staking.OwnerQuery{
+		Height: consensus.HeightLatest,
+		Owner:  pubKey,
+	})
+	if err != nil {
+		return 0, fmt.Errorf("stakingClient.AccountInfo %s: %w", pubKey, err)
+	}
+	return account.General.Nonce, nil
+}
+
+func (registration) Run(gracefulExit context.Context, rng *rand.Rand, conn *grpc.ClientConn, cnsc consensus.ClientBackend, rtc runtimeClient.RuntimeClient) error {
+	ctx := context.Background()
+	var err error
+
+	if err = ns.UnmarshalHex("0000000000000000000000000000000000000000000000000000000000000002"); err != nil {
+		panic(err)
+	}
+
+	baseDir := viper.GetString(cmdCommon.CfgDataDir)
+	nodeIdentitiesDir := filepath.Join(baseDir, "node-identities")
+	if err = common.Mkdir(nodeIdentitiesDir); err != nil {
+		return fmt.Errorf("txsource/registry: failed to create node-identities dir: %w", err)
+	}
+
+	// Fetch fees.
+	// TODO: Don't dump everything, instead add a method to query just the parameters.
+	genesisDoc, err := cnsc.StateToGenesis(ctx, 1)
+	if err != nil {
+		return fmt.Errorf("failed to query state at genesis: %w", err)
+	}
+	// Entity registration fees.
+	entityRegisterFee := transaction.Fee{
```

Comment:
Would be cool if this used `EstimateGas` (currently not exposed in the consensus `ClientBackend`, but should be trivial to do), related to #2723 (just the API part) instead of manually calculating the gas cost.

Reply:
cool! done and works :+1:

Reply:
Nice!